File: figs/ser_contour/gen_ser_contour.py
# Copyright 2013-2014 Matthias Wilhelm

# This program is free software: you can redistribute it and/or modify
# it under the terms of the GNU General Public License as published by
# the Free Software Foundation, either version 3 of the License, or
# (at your option) any later version.

# This program is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
# GNU General Public License for more details.

# You should have received a copy of the GNU General Public License
# along with this program.  If not, see <http://www.gnu.org/licenses/>.

from __future__ import print_function

import argparse
import os
import sys

sys.path.append("../../src")

import numpy as np
import phitau_opt as pt
import tools

######################################################
#################      SETTINGS      #################
######################################################
T = 1.0
nsteps = 256

settings = {
	"As":     1.0,
	"Au":     np.sqrt(1e4),
	"nsyms":  256,
	"nsteps": nsteps,

	"tau_range": np.linspace(-1.5*T, 1.5*T, num=nsteps),
	"phi_range": np.linspace(-np.pi, np.pi, num=nsteps)
}

globals().update(settings)
######################################################
######################################################

def do_gen(content, decision):
	SER_S = np.empty((tau_range.shape[0], phi_range.shape[0]))
	SER_U = np.empty((tau_range.shape[0], phi_range.shape[0]))

	for tau_idx, tau in enumerate(tau_range):
		print("tau = %.2f" % (tau))

		# symbols of (synch"ed, unsynch"ed) sender
		if content in ("same",):
			tmp_syms = np.random.randint(16, size=nsyms+2)
			send_syms = [tmp_syms, tmp_syms]
		else:
			send_syms = np.random.randint(16, size=2*(nsyms+2)).reshape(2, nsyms+2)

		send_syms_s, send_syms_u = send_syms[0][1:-1], send_syms[1][1:-1]
		send_chips = pt.map_chips(*send_syms)

		RECV_CHIPS_I = pt.detect_i(send_chips[:2], send_chips[2:], phi_range, tau, As, Au)
		RECV_CHIPS_Q = pt.detect_q(send_chips[:2], send_chips[2:], phi_range, tau, As, Au)

		for phi_idx in range(len(phi_range)):
			recv_chips       = np.empty(2*RECV_CHIPS_I.shape[0])
			recv_chips[ ::2] = RECV_CHIPS_I[:,phi_idx]
			recv_chips[1::2] = RECV_CHIPS_Q[:,phi_idx]

			# slice bits to simulate hard decision decoder
			if decision in ("hard",):
				recv_chips = np.sign(recv_chips)

			recv_syms = pt.detect_syms_corr(recv_chips)[1:-1]

			SER_S[tau_idx, phi_idx] = sum(recv_syms != send_syms_s) / (1.0*len(recv_syms))
			SER_U[tau_idx, phi_idx] = sum(recv_syms != send_syms_u) / (1.0*len(recv_syms))

	np.savez_compressed("data/ser_Au%.2f_%s_%s_v2.npz"%(Au, content, decision),
                        SER_S=SER_S, SER_U=SER_U, **settings)

if __name__ == "__main__":
	argp = argparse.ArgumentParser()
	argp.add_argument("content",  choices=("same", "unif"), help="Relation between data content in the two transmitted packets")
	argp.add_argument("decision", choices=("soft", "hard"), help="Bit decision for DSSS decoding (SDD, HDD)")

	args = argp.parse_args()

	if not tools.overwrite_ok("data/ser_Au%.2f_%s_%s_v2.npz"%(Au, args.content, args.decision)):
		sys.exit(0)

	do_gen(args.content, args.decision)

File: julia/figs/ser_contour/gen_ser_contour.jl
# Copyright 2015 Matthias Wilhelm

# This program is free software: you can redistribute it and/or modify
# it under the terms of the GNU General Public License as published by
# the Free Software Foundation, either version 3 of the License, or
# (at your option) any later version.

# This program is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
# GNU General Public License for more details.

# You should have received a copy of the GNU General Public License
# along with this program.  If not, see <http://www.gnu.org/licenses/>.

import NPZ

push!(LOAD_PATH, "../../src")
import PhiTau
@everywhere const pt = PhiTau

@everywhere using DistributedArrays

# -----------------------------------------------------------------------------

function main()
	if length(ARGS) == 2
		do_gen(ARGS...)
	else
		println("Two required arguments: content={same,unif}, decision={hard,soft}")
	end
end

function main_vid()
	if length(ARGS) == 2
		for SIR in -30:10
			do_gen(ARGS...; SIR=SIR)
		end
	else
		println("Two required arguments: content={same,unif}, decision={hard,soft}")
	end
end

# -----------------------------------------------------------------------------

function do_gen(content::String, decision::String; SIR=-20.0)
	const T  = 1.0
	const Au = √(10^(-SIR/10))
	const nsyms  = 2^12
	const nsteps = 2^9

	const τ_range = linspace(-2.5T, 2.5T, nsteps)
	const 𝜑_range = linspace( -π,  π, nsteps)

	SER_S = dzeros(length(𝜑_range), length(τ_range))
	SER_U = dzeros(length(𝜑_range), length(τ_range))

	@sync @parallel for i=1:length(workers())
		calcSER!(SER_S, SER_U, content, decision, τ_range, 𝜑_range, As=1.0, Au=Au, nsyms=nsyms)
	end

	NPZ.npzwrite("data/serc_$(content)_$(decision)_SIR_$(round(Int, SIR)).npz",
				 Dict("SER_S"=>convert(Array, SER_S), "SER_U"=>convert(Array, SER_U),
					  "tau_range"=>τ_range, "phi_range"=>𝜑_range,
					  "As"=>1.0, "Au"=>Au, "nsyms"=>nsyms, "nsteps"=>nsteps))
end

# -----------------------------------------------------------------------------

@everywhere function calcSER!(dSER_S::DArray, dSER_U::DArray, content::String, decision::String,
		τ_range::Vector{Float64}, 𝜑_range::Vector{Float64}; As=1.0, Au=√100, nsyms=2^10, nsteps=2^8)

	if !(content in ("same", "unif"))
		throw(ArgumentError("Content must be in {same, unif}"))
	end
	if !(decision in ("hard", "soft"))
		throw(ArgumentError("Decision must be in {hard, soft}"))
	end

	SER_S = localpart(dSER_S)
	SER_U = localpart(dSER_U)

	lidx = localindexes(dSER_U)

	α_send_syms = zeros(Int, nsyms)
	β_send_syms = zeros(Int, nsyms)
	α_send_chips = zeros(Complex128, 16*nsyms)
	β_send_chips = zeros(Complex128, 16*nsyms)

	recv_syms = zeros(Int, nsyms)


	for (τ_idx, τ) in enumerate(τ_range[lidx[2]])
		println("τ = ", @sprintf("% .3f", τ), ". Worker progress: ", @sprintf("%6.2f", 100.τ_idx/length(lidx[2])), "%")

		if content == "same"
			rand!(α_send_syms, 1:16)
			β_send_syms  = α_send_syms
			pt.map_chips!(α_send_chips, α_send_syms)
			β_send_chips = α_send_chips
		else
			rand!(α_send_syms, 1:16)
			rand!(β_send_syms, 1:16)
			pt.map_chips!(α_send_chips, α_send_syms)
			pt.map_chips!(β_send_chips, β_send_syms)
		end

		RECV_CHIPS = As*α_send_chips .+ Au*pt.Λu(β_send_chips, 𝜑_range[lidx[1]], τ)

		if decision == "hard"
			RECV_CHIPS = complex(sign(real(RECV_CHIPS)), sign(imag(RECV_CHIPS)))
		end

		tic()
		for 𝜑_idx in 1:length(lidx[1])
			pt.detect_syms_corr!(recv_syms, RECV_CHIPS[:,𝜑_idx])

			SER_S[𝜑_idx,τ_idx] = countnz(recv_syms .≠ α_send_syms)/nsyms
			SER_U[𝜑_idx,τ_idx] = countnz(recv_syms .≠ β_send_syms)/nsyms
		end
		toc()
	end
end

# -----------------------------------------------------------------------------

main_vid()
